### PR TITLE
fix(storage): Update api for bucket update

### PR DESCRIPTION
**Pull request by SandeepTuniki**
closes: http://b/232352080 (internal)

This PR changes the following implementation:
```
service method "update_bucket" --calls--> apiary method "patch_bucket"
```
to:
```
service method "update_bucket" --calls--> apiary method "update_bucket"
```

It also provides a retry option for the `update_bucket` method and sets it to `0` as default value.

### CODE REVIEW

**Review by bajajneha27 — APPROVED** [review]
LGTM

File: google-cloud-storage/lib/google/cloud/storage/service.rb [hunk 1]
```diff
@@ -721,6 +721,35 @@ def update_hmac_key access_id, hmac_key_metadata_object,
...
+        def update_bucket bucket_name,
+                          bucket_gapi = nil,
+                          predefined_acl: nil,
+                          predefined_default_acl: nil,
+                          if_metageneration_match: nil,
+                          if_metageneration_not_match: nil,
+                          user_project: nil,
+                          options: {}
+          bucket_gapi ||= Google::Apis::StorageV1::Bucket.new
+          bucket_gapi.acl = [] if predefined_acl
```
Comment (NivedhaSenthil) [inline google-cloud-storage/lib/google/cloud/storage/service.rb:735]: why are we setting these to empty array ? 
predefined_acl and predefined_default_acl options seems to be not used, should they be passed on to update_bucket ?

Comment (SandeepTuniki) [reply google-cloud-storage/lib/google/cloud/storage/service.rb:735]: > why are we setting these to empty array ?

 I was replicating the functionality of `patch_bucket`, so I kept the code identical. I'm not sure of the intent.
https://github.com/googleapis/google-cloud-ruby/blob/a39199b1a995d3bd32e214f5c3a8d3863aa0cc0e/google-cloud-storage/lib/google/cloud/storage/service.rb#L126-L136

Comment (SandeepTuniki) [reply google-cloud-storage/lib/google/cloud/storage/service.rb:735]: > predefined_acl and predefined_default_acl options seems to be not used, should they be passed on to update_bucket ?

Yes, I missed this. Thanks for catching it. I changed the code to pass it to apiary's update_bucket.

**Review by NivedhaSenthil — APPROVED** [review]

File: google-cloud-storage/acceptance/storage/bucket_test.rb [hunk 2]
```diff
@@ -61,7 +61,7 @@
     _(one_off_bucket.website_404).must_equal "not_found.html"
     _(one_off_bucket.requester_pays).must_equal true
     # labels with symbols are not strings
-    _(one_off_bucket.labels).must_equal({ "foo" => "bar" })
+    _(one_off_bucket.labels).must_equal({ :foo => :bar })
```
Comment (SandeepTuniki) [inline google-cloud-storage/acceptance/storage/bucket_test.rb:64]: This line is - strictly speaking - not part of this PR's functionality. This test has been failing sporadically for the past few months (though the build has been reporting the error in line 42).

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 3]
```diff
@@ -1169,10 +1169,22 @@ def update if_metageneration_match: nil, if_metageneration_not_match: nil
           updater.check_for_changed_labels!
           updater.check_for_mutable_cors!
           updater.check_for_mutable_lifecycle!
-          return if updater.updates.empty?
-          patch_gapi! updater.updates,
-                      if_metageneration_match: if_metageneration_match,
-                      if_metageneration_not_match: if_metageneration_not_match
+
+          attributes = Array(updater.updates)
```
Comment (dazuma) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:1173]: Consider creating a helper method `update_gapi!` to parallel the existing `patch_gapi!`. That will make it easier to see how this code relates to the existing code, clarifying things for potential future refactors.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 4]
```diff
@@ -1169,10 +1169,22 @@ def update if_metageneration_match: nil, if_metageneration_not_match: nil
...
+          attributes.flatten!
+          return if attributes.empty?
+          ensure_service!
+          update_args = Hash[attributes.map do |attr|
+            [attr, @gapi.send(attr)]
+          end]
+          update_gapi = API::Bucket.new(**update_args)
+          service.update_bucket name,
```
Comment (dazuma) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:1181]: In particular, I notice that `patch_gapi!` sets the `@gapi` instance variable here, whereas your code does not. Is that intentional, and why is it different?

Comment (SandeepTuniki) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb]: I missed this. Thanks for spotting it. I added it now.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 3 above)
Comment (SandeepTuniki) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb]: Thanks @dazuma. I added the helper method. Now the flow of the code is:
```
bucket.update --> bucket.update_gapi! --> service.update_bucket --> apiary.update_bucket
```

File: google-cloud-storage/acceptance/storage/bucket_test.rb (hunk 2 above)
Comment (SandeepTuniki) [reply google-cloud-storage/acceptance/storage/bucket_test.rb]: This spec is still flaky & is failing. Reverting this change made it pass. I'll investigate it in a separate PR.

**Comment by SandeepTuniki** [comment]
The failing build is due to flaky specs and are unrelated to this PR. They're being attempted to fix at #19147

**Pull request merged**